### PR TITLE
Cstor51

### PR DESCRIPTION
**Pull request by mynktl**
- **API to set "rebuild mode" to volume during rebuild**
       - uzfs_set_rebuilding_mode (If replica is in downgraded mode then this API should be called to set write only mode, which will start to populate incoming_io_tree with normal write io and incoming rebuild io will be checked and write accordingly.)
       - uzfs_unset_rebuilding_mode (To unset write only mode once rebuild finished)

- **Changes in uzfs_write_data API to handle rebuild io.**
Added extra argument in uzfs_write_data API.
uzfs_write_data(void *zv, char *buf, uint64_t offset, uint64_t len, void *md, boolean_t is_rebuild);
is_rebuild should be set to TRUE, if uzfs_write_data is called for rebuild purpose else should be set to FALSE (default mode).

- **Test case to verify rebuild with two spa/zvol.(only uzfs_write_data API)**

### CODE REVIEW

File: lib/libzpool/uzfs_io.c
Comment (satbirchhikara) [inline lib/libzpool/uzfs_io.c]: Can we have else for readability ?

File: cmd/uzfs_test/uzfs_txg_diff.c
Comment (vishnuitta) [inline cmd/uzfs_test/uzfs_txg_diff.c]: uzfs_test_txg_diff_traverse_cb

File: cmd/uzfs_test/uzfs_rebuilding.c
Comment (jkryl) [inline cmd/uzfs_test/uzfs_rebuilding.c:279]: we could release the mutex right before this if condition which would allow producer to enqueue more items to the list while this thread does something else. And then acquire it back at the end of loop. Not sure if it is worth of it. but just something we can consider to allow greater parallelism.

File: cmd/uzfs_test/uzfs_rebuilding.c
Comment (jkryl) [inline cmd/uzfs_test/uzfs_rebuilding.c]: double assignment to start_txg. The same problem below with end_txg. Is it ok if start_txg is always 0?

File: include/sys/uzfs_zvol.h
Comment (vishnuitta) [inline include/sys/uzfs_zvol.h]: push these rebuild related things into some data structure

File: include/sys/uzfs_zvol.h
Comment (vishnuitta) [inline include/sys/uzfs_zvol.h]: some comments

File: cmd/uzfs_test/uzfs_rebuilding.c
Comment (jkryl) [inline cmd/uzfs_test/uzfs_rebuilding.c:677]: Looks nice, but in order to be truly useful any kind of error encountered during the test should result in failure and program's return code != 0. Currently an error message is printed, however for test automation that is not enough. Ideally even the write or read zvol errors should terminate the test and return failure.

File: include/uzfs_io.h
Comment (vishnuitta) [inline include/uzfs_io.h]: if IO is from target, it will be set to FALSE, else, it should TRUE

File: include/uzfs_io.h
Comment (vishnuitta) [inline include/uzfs_io.h]: pass the value either TRUE or FALSE to set, and have a get function

File: lib/libzpool/uzfs_io.c
Comment (jkryl) [inline lib/libzpool/uzfs_io.c]: "else" would be nicer here :)

File: lib/libzpool/uzfs_io.c
Comment (jkryl) [inline lib/libzpool/uzfs_io.c]: could you share more information about how we deal with synchronization between setting rebuild mode and IO path? Do we need any mutex here?

File: lib/libzpool/uzfs_io.c
Comment (satbirchhikara) [inline lib/libzpool/uzfs_io.c]: len_in_first_aligned_block is getting set to zero if count is non-zero. I think we need to handle it in cases where we have overlapping writes.

File: lib/libzpool/uzfs_io.c
Comment (satbirchhikara) [inline lib/libzpool/uzfs_io.c]: This comment is just to improve readability. Its upto you...take it or ignore it.
<Code snippet>
        if (zv->in_rebuilding_mode) {
                if (!is_rebuild) {
                        uzfs_add_to_incoming_io_tree(zv, offset, len);
                } else {
                        count = uzfs_search_incoming_io_tree(zv, offset,
                            len, (void **)&chunk_io);
                }
        }

chunk_io:
        if (count) {
                node = list_remove_head(chunk_io);
                offset = node->offset;
                len = node->len;
                end = offset + len;
                umem_free(node, sizeof (*node));
                wrote = offset - orig_offset;
                len_in_first_aligned_block = 0;
                count--;
                zv->rebuild_bytes += len;
        }
</Code snippet>

File: lib/libzpool/uzfs_mtree.c
Comment (satbirchhikara) [inline lib/libzpool/uzfs_mtree.c]: As per my understanding kmem_alloc is the right interface.

File: lib/libzpool/uzfs_io.c
Comment (satbirchhikara) [inline lib/libzpool/uzfs_io.c]: I think here too we need to use kmem_free.

File: lib/libzpool/uzfs_io.c
Comment (satbirchhikara) [inline lib/libzpool/uzfs_io.c]: Suggestion@ We can make it little generic by passing enum to set a particular state on zv.

File: lib/libzpool/uzfs_mtree.c
Comment (jkryl) [inline lib/libzpool/uzfs_mtree.c]: I got quite confused by name of this function, thinking that search_incoming_io_tree would return an entry from the tree which overlaps with offset, len parameters. But in fact it returns all parts from the chunk from tree which are not overlapping. I'm not sure if we can find a better name. If not perhaps a comment explaining this could help.

File: lib/libzpool/uzfs_mgmt.c
Comment (satbirchhikara) [inline lib/libzpool/uzfs_mgmt.c]: These can be created/initialised  when replica receive START_REBUILD cmd from iSCSI controller.

File: lib/libzpool/uzfs_mgmt.c
Comment (satbirchhikara) [inline lib/libzpool/uzfs_mgmt.c]: Same comment goes here too, we can remove/destroy when we done with Rebuild.

**Review by satbirchhikara — COMMENTED** [review]
Hi Mayank, I have looked into code change. It is in a good state. Few minor comments has been posted here.

File: lib/libzpool/uzfs_mtree.c
Comment (vishnuitta) [inline lib/libzpool/uzfs_mtree.c]: incoming_io_tree should be protected with lock

File: lib/libzpool/uzfs_mtree.c
Comment (vishnuitta) [inline lib/libzpool/uzfs_mtree.c]: put comments like (b_entry->offset, b_end, offset, a_end)

File: lib/libzpool/uzfs_mtree.c
Comment (vishnuitta) [inline lib/libzpool/uzfs_mtree.c]: (b_entry->offset, offset, a_end, b_end)

File: lib/libzpool/uzfs_mtree.c
Comment (vishnuitta) [inline lib/libzpool/uzfs_mtree.c]: (b_entry->offset, offset, b_end, a_end).. actually, 'if' condition is not required.

File: lib/libzpool/uzfs_mtree.c
Comment (vishnuitta) [inline lib/libzpool/uzfs_mtree.c]: (offset, b_end, a_entry->offset, a_end)

File: lib/libzpool/uzfs_mtree.c
Comment (vishnuitta) [inline lib/libzpool/uzfs_mtree.c]: (offset, a_entry->offset, b_end, a_end)

File: lib/libzpool/uzfs_mtree.c
Comment (vishnuitta) [inline lib/libzpool/uzfs_mtree.c]: (offset, a_entry->offset, a_end, b_end).. same here, 'if' case not required

File: lib/libzpool/uzfs_io.c
Comment (vishnuitta) [reply lib/libzpool/uzfs_io.c]: yes.. setting this to 0 is not correct.. it should be:
```
        r_offset1 = (offset / blocksize) * blocksize;
 	len_in_first_aligned_block = (blocksize - (offset - r_offset1));
 
 	if (len_in_first_aligned_block > len)
                len_in_first_aligned_block = len;
```

**Review by vishnuitta — COMMENTED** [review]
Putting this not to forget:
- how reads are handled in case of partial writes during degraded mode?

File: lib/libzpool/uzfs_io.c
Comment (vishnuitta) [reply lib/libzpool/uzfs_io.c]: but, with len_in_first_aligned_block being 0 also, things works fine. This variable makes sure that IOs are aligned with block size even if initial offset is misaligned to blocksize.

File: lib/libzpool/uzfs_io.c
Comment (vishnuitta) [reply lib/libzpool/uzfs_io.c]: better do list_destroy also

File: cmd/uzfs_test/uzfs_txg_diff.c
Comment (mynktl) [reply cmd/uzfs_test/uzfs_txg_diff.c]: ok

File: include/sys/uzfs_zvol.h
Comment (mynktl) [reply include/sys/uzfs_zvol.h]: ok

File: include/sys/uzfs_zvol.h
Comment (mynktl) [reply include/sys/uzfs_zvol.h]: ok

File: include/uzfs_io.h
Comment (mynktl) [reply include/uzfs_io.h]: ok

File: include/uzfs_io.h
Comment (mynktl) [reply include/uzfs_io.h]: ok

File: cmd/uzfs_test/uzfs_rebuilding.c
Comment (mynktl) [reply cmd/uzfs_test/uzfs_rebuilding.c]: that i have added for logging purpose to check start_txg and end_txg. I will remove unnecessary assignment.
First i took start_txg of replica which needs to be rebuild but that didn't work out as both replica will never write with same transaction number. So, i took start_txg as 0. This is not a proper workaround but for testing purpose i think it is ok.

File: cmd/uzfs_test/uzfs_rebuilding.c
Comment (mynktl) [reply cmd/uzfs_test/uzfs_rebuilding.c:677]: @jkryl  it is already there for read case. In replica_reader_thread if there is any mismatch in data between two replica then it will print errors and exit with 1.
For write purpose, i will add.

File: lib/libzpool/uzfs_io.c
Comment (mynktl) [reply lib/libzpool/uzfs_io.c]: ok :)

File: lib/libzpool/uzfs_io.c
Comment (mynktl) [reply lib/libzpool/uzfs_io.c]: synchronization between setting rebuild mode and IO path should be handled by caller API only. Since IO path can be parallel so i have avoided any synchronization at this level.

File: lib/libzpool/uzfs_mtree.c
Comment (mynktl) [reply lib/libzpool/uzfs_mtree.c]: ok.. Comment is already there in header file. I will add it to c file also :)

File: lib/libzpool/uzfs_io.c
Comment (mynktl) [reply lib/libzpool/uzfs_io.c]: ok

File: lib/libzpool/uzfs_io.c
Comment (mynktl) [reply lib/libzpool/uzfs_io.c]: ok. I will make changes accordingly.

File: lib/libzpool/uzfs_mtree.c
Comment (mynktl) [reply lib/libzpool/uzfs_mtree.c]: I think umem_ related API's are for userspace memory operations. So I used umem_* API for memory allocation/free.

File: lib/libzpool/uzfs_io.c
Comment (mynktl) [reply lib/libzpool/uzfs_io.c]: ok

File: lib/libzpool/uzfs_io.c
Comment (mynktl) [reply lib/libzpool/uzfs_io.c]: ok

File: lib/libzpool/uzfs_mgmt.c
Comment (mynktl) [reply lib/libzpool/uzfs_mgmt.c]: yes. Once we have that flow, we can shift this initialization there. As of now, that part is not yet implemented, so for working flow only i kept it here.

File: lib/libzpool/uzfs_mgmt.c
Comment (mynktl) [reply lib/libzpool/uzfs_mgmt.c]: yes. We can shift this operations to somewhere during integration or complete flow. As of now, i have kept this operation here just to make it complete flow without replica/target. I will add comments here.

File: lib/libzpool/uzfs_mtree.c
Comment (mynktl) [reply lib/libzpool/uzfs_mtree.c]: ok

File: lib/libzpool/uzfs_mtree.c
Comment (mynktl) [reply lib/libzpool/uzfs_mtree.c]: ok

File: lib/libzpool/uzfs_mtree.c
Comment (mynktl) [reply lib/libzpool/uzfs_mtree.c]: ok

File: lib/libzpool/uzfs_mtree.c
Comment (mynktl) [reply lib/libzpool/uzfs_mtree.c]: I think it is required, 
if b_entry --- offset:100 and len:100
and search_entry --- offset:130 and len:200
then it will become (100, 130, 200, 330), to remove overlapping of 130 to 200, we need this check, right?
I will also add unit test for all this possibility. since rebuild test is not enough to check all scenario. What you suggest?

File: lib/libzpool/uzfs_mtree.c
Comment (mynktl) [reply lib/libzpool/uzfs_mtree.c]: ok

File: cmd/uzfs_test/uzfs_rebuilding.c
Comment (jkryl) [reply cmd/uzfs_test/uzfs_rebuilding.c:677]: thx. Write and verify cases are missing this as of now.

File: lib/libzpool/uzfs_mtree.c
Comment (jkryl) [reply lib/libzpool/uzfs_mtree.c]: the mappings are as follows: kmem -> umem -> malloc :) So in theory we could use any of these three, however for consistency reasons with the rest of the zfs code, I would recommend only the first two.

File: lib/libzpool/uzfs_io.c
Comment (mynktl) [reply lib/libzpool/uzfs_io.c]: uzfs_search_incoming_io_tree may return 0 if incoming_io_tree is NULL. This value is checked further  so i kept this code snippet under if (zv->in_rebuilding_mode) if (is_rebuild) only.
It also makes uzfs_write_data less messy by keeping rebuilding code in few blocks only.

File: lib/libzpool/uzfs_mtree.c
Comment (mynktl) [reply lib/libzpool/uzfs_mtree.c]: thats right @jkryl . if we compile it for userspace then kmem and umem will lead to malloc only. But if code is meant for userspace then i think its better to use umem_ API.

File: lib/libzpool/uzfs_mtree.c
Comment (mynktl) [reply lib/libzpool/uzfs_mtree.c]: ok

File: lib/libzpool/uzfs_mtree.c
Comment (mynktl) [reply lib/libzpool/uzfs_mtree.c]: ok

File: cmd/uzfs_test/uzfs_rebuilding.c
Comment (mynktl) [reply cmd/uzfs_test/uzfs_rebuilding.c:279]: yes. Thats right. We don't need to hold lock once we dequeue entry from list. I will add this change.

File: cmd/uzfs_test/uzfs_rebuilding.c
Comment (mynktl) [reply cmd/uzfs_test/uzfs_rebuilding.c:677]: for verify, we don't need. VERIFY will do abort in false case. right?

File: lib/libzpool/uzfs_io.c
Comment (mynktl) [reply lib/libzpool/uzfs_io.c]: nice suggestion. I will make it generic.

**Comment by mynktl** [comment]
I have addressed review comments at my best.
Regarding read in case of partial writes, i will add this cases in my next issue.
Please let me know if i have missed any change from review comments or any necessary changes still required.

**Review by jkryl — APPROVED** [review]
thanks mayank

File: include/sys/uzfs_zvol.h
Comment (vishnuitta) [inline include/sys/uzfs_zvol.h]: REBUILDING state should be part of zvol_rebuild_status rather than zvol_status

File: lib/libzpool/uzfs_io.c
Comment (vishnuitta) [inline lib/libzpool/uzfs_io.c]: check for ret as 0 also

File: lib/libzpool/uzfs_io.c
Comment (vishnuitta) [inline lib/libzpool/uzfs_io.c]: Few comments here:
- this need to be DEGRADED state, so that, we add right from beginning of replica
- function became too big and can be reduced by some macro
- how about using a loop here?

File: lib/libzpool/uzfs_mgmt.c
Comment (vishnuitta) [inline lib/libzpool/uzfs_mgmt.c]: on boot, we are supposed to mark zv as DEGRADED

Comment (mynktl) [reply lib/libzpool/uzfs_mgmt.c]: ok

File: cmd/uzfs_test/uzfs_rebuilding.c
Comment (vishnuitta) [inline cmd/uzfs_test/uzfs_rebuilding.c]: replica_start_time and replica_down_time are good.. we can have rebuild_start_time also similarly.. we shouldn't assume that rebuild will start at the same time when replica comes up

File: include/sys/uzfs_zvol.h
Comment (mynktl) [reply include/sys/uzfs_zvol.h]: ok

File: lib/libzpool/uzfs_io.c
Comment (mynktl) [reply lib/libzpool/uzfs_io.c]: ok

File: cmd/uzfs_test/uzfs_rebuilding.c
Comment (mynktl) [reply cmd/uzfs_test/uzfs_rebuilding.c]: ok

**Review by jkryl — APPROVED** [review]

**Review by vishnuitta — APPROVED** [review]
changes are good.
let us initialize rebuild_status to state like NOT_STARTED in next commit

**Comment by mynktl** [comment]
@vishnuitta  We can merge this branch as changes after @jkryl approval were reviewed by @jkryl.
@jkryl, Please correct me if i am wrong with above statement.

**Pull request merged**
